numpy/_core/src/umath/string_buffer.h

```diff
@@ -391,5 +587,537 @@ operator>=(Buffer<enc> lhs, Buffer<enc> rhs)
+            case ENCODING::UTF8:
+            {
+                if (num_bytes_for_utf8_character((const unsigned char *)(buf1 + start).buf) > 1) {
+                    goto multibyte_search;
+                }
+                // fall through to the ASCII case because this is a one-byte character
+            }
+            case ENCODING::ASCII:
+            {
+                char ch = *buf2;
+                CheckedIndexer<char> ind((buf1 + start).buf, end - start);
+                result = (npy_intp) rfindchar(ind, end - start, ch);
+                break;
+            }
+            case ENCODING::UTF32:
+            {
+                npy_ucs4 ch = *buf2;
+                CheckedIndexer<npy_ucs4> ind((npy_ucs4 *)(buf1 + start).buf, end - start);
+                result = (npy_intp) rfindchar(ind, end - start, ch);
+                break;
+            }
+        }
+        if (result == -1) {
+            return (npy_intp) -1;
+        }
+        else {
+            return result + (npy_intp) start;
+        }
+    }
+
+  multibyte_search:
+
+    npy_intp pos;
+    switch (enc) {
+        case ENCODING::ASCII:
+            pos = (npy_intp) fastsearch((buf1 + start).buf, end - start,
+                                        buf2.buf, len2, -1, FAST_RSEARCH);
+            break;
+        case ENCODING::UTF8:
+        {
+            char *start_loc = NULL;
+            char *end_loc = NULL;
+            find_start_end_locs(buf1.buf, (buf1.after - buf1.buf), start, end, &start_loc, &end_loc);
+            pos = fastsearch(start_loc, end_loc - start_loc, buf2.buf, buf2.after - buf2.buf, -1, FAST_RSEARCH);
+            // pos is the byte index, but we need the character index
+            if (pos > 0) {
+                pos = utf8_character_index(start_loc, start_loc - buf1.buf, start, pos, buf1.after - start_loc);
+            }
+            break;
+        }
+        case ENCODING::UTF32:
+            pos = (npy_intp) fastsearch((npy_ucs4 *)(buf1 + start).buf, end - start,
+                                        (npy_ucs4 *)buf2.buf, len2, -1, FAST_RSEARCH);
+            break;
+    }
+    if (pos >= 0) {
+        pos += start;
+    }
+    return pos;
+}
+
+
+/*
+ * Count the number of occurences of buf2 in buf1 between
+ * start (inclusive) and end (exclusive)
+ */
+template <ENCODING enc>
+static inline npy_intp
+string_count(Buffer<enc> buf1, Buffer<enc> buf2, npy_int64 start, npy_int64 end)
+{
+    size_t len1 = buf1.num_codepoints();
+    size_t len2 = buf2.num_codepoints();
+
+    adjust_offsets(&start, &end, len1);
+    if (end < start || end - start < static_cast<npy_int64>(len2)) {
+        return (npy_intp) 0;
+    }
+
+    if (len2 == 0) {
+        return (end - start) < PY_SSIZE_T_MAX ? end - start + 1 : PY_SSIZE_T_MAX;
+    }
+
+    npy_intp count;
+    switch (enc) {
+        case ENCODING::UTF8:
+        case ENCODING::ASCII:
+            count = (npy_intp) fastsearch((buf1 + start).buf, end - start, buf2.buf, len2,
```

Comment:
It seems a bit counter-intuitive that `find` and `rfind` need a different case for `UTF8` that does a bit of preprocessing for what bytes to pass to `fastsearch`, but `count` doesn't. Is that really so?

Reply:
Yes, that is so. 

`count` is given the exact byte pattern to match and doesn't care about character indices in either the haystack or needle, so we can just pass those UTF8 bytes directly to `fastsearch`. We also don't care that fastsearch returned a byte index, since we only care about the number of matches.

For `find` and `rfind`, the API is defined terms of characters, so I need to advance by `start` characters and go backward from the end of the string by `end` characters to find the correct byte offsets. Additionally, the result is defined in terms of a character index, so I also need to postprocess to go from the byte index returned by `fastsearch` to get the character index the python API expects. No matter what for UTF8, there's going to be some pre and post-processing

Does that make sense?